File: Dhiva/moviecatalogue/app/src/main/java/id/dhiva/moviecatalogue/viewmodel/ViewModelFactory.java
package id.dhiva.moviecatalogue.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import id.dhiva.moviecatalogue.data.DataRepository;
import id.dhiva.moviecatalogue.data.di.Injection;
import id.dhiva.moviecatalogue.ui.detail.DetailViewModel;
import id.dhiva.moviecatalogue.ui.favorite.favmovie.FavMovieViewModel;
import id.dhiva.moviecatalogue.ui.favorite.favtv.FavTvViewModel;
import id.dhiva.moviecatalogue.ui.movie.MovieViewModel;
import id.dhiva.moviecatalogue.ui.tv.TvViewModel;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private static volatile ViewModelFactory INSTANCE;
    private final DataRepository dataRepository;

    private ViewModelFactory(DataRepository dataRepository){
        this.dataRepository = dataRepository;
    }

    public static ViewModelFactory getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (ViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ViewModelFactory(Injection.provideRepository(application));
                }
            }
        }
        return INSTANCE;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass.isAssignableFrom(MovieViewModel.class)) {
            //noinspection unchecked
            return (T) new MovieViewModel(dataRepository);
        } else if (modelClass.isAssignableFrom(DetailViewModel.class)) {
            //noinspection unchecked
            return (T) new DetailViewModel(dataRepository);
        } else if (modelClass.isAssignableFrom(TvViewModel.class)) {
            //noinspection unchecked
            return (T) new TvViewModel(dataRepository);
        } else if(modelClass.isAssignableFrom(FavMovieViewModel.class)){
            //noinspection unchecked
            return (T) new FavMovieViewModel(dataRepository);
        } else if(modelClass.isAssignableFrom(FavTvViewModel.class)){
            //noinspection unchecked
            return (T) new FavTvViewModel(dataRepository);
        }

        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: Submission2/app/src/main/java/cikal/dicoding/submission2/viewmodel/ViewModelFactory.kt
package cikal.dicoding.submission2.viewmodel

import cikal.dicoding.submission2.data.source.CatalogueRepository
import androidx.lifecycle.ViewModelProvider
import android.content.Context
import androidx.lifecycle.ViewModel
import cikal.dicoding.submission2.di.Injection
import cikal.dicoding.submission2.ui.detail.DetailCatalogueViewModel
import cikal.dicoding.submission2.ui.movie.MovieViewModel
import cikal.dicoding.submission2.ui.tvshow.TvShowViewModel

class ViewModelFactory private constructor(private val mCatalogueRepository: CatalogueRepository) :
    ViewModelProvider.NewInstanceFactory() {

    companion object {
        @Volatile
        private var instance: ViewModelFactory? = null

        fun getInstance(context: Context): ViewModelFactory =
            instance ?: synchronized(this) {
                instance ?: ViewModelFactory(Injection.provideRepository(context))
            }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return when {
            modelClass.isAssignableFrom(MovieViewModel::class.java) -> {
                MovieViewModel(mCatalogueRepository) as T
            }
            modelClass.isAssignableFrom(TvShowViewModel::class.java) -> {
                TvShowViewModel(mCatalogueRepository) as T
            }
            modelClass.isAssignableFrom(DetailCatalogueViewModel::class.java) -> {
                DetailCatalogueViewModel(mCatalogueRepository) as T
            }
            else -> throw Throwable("Unknown ViewModel class: " + modelClass.name)
        }
    }
}